### PR TITLE
Balder/same element in backend

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe and @arnej27959 PR

### CODE REVIEW

File: searchcorespi/src/vespa/searchcorespi/index/indexcollection.cpp [hunk 1]
```diff
@@ -179,28 +179,29 @@ class CreateBlueprintVisitor : public search::query::QueryVisitor {
...
+    void visit(Equiv &)   override { }
+    void visit(Rank &)    override { }
+    void visit(Near &)    override { }
+    void visit(ONear &)   override { }
+
+    void visit(WeightedSetTerm &n) override { visitTerm(n); }
+    void visit(DotProduct &n)      override { visitTerm(n); }
+    void visit(WandTerm &n)        override { visitTerm(n); }
+    void visit(Phrase &n)          override { visitTerm(n); }
+    void visit(SameElement &n)     override { visitTerm(n); }
```
Comment (havardpe) [inline searchcorespi/src/vespa/searchcorespi/index/indexcollection.cpp]: should not be handled as a term in this context

File: searchcore/src/vespa/searchcore/proton/matching/blueprintbuilder.cpp [hunk 2]
```diff
@@ -124,29 +123,30 @@ class BlueprintBuilderVisitor :
...
+    void visit(ProtonRank &n)    override { buildIntermediate(new RankBlueprint(), n); }
+    void visit(ProtonNear &n)    override { buildIntermediate(new NearBlueprint(n.getDistance()), n); }
+    void visit(ProtonONear &n)   override { buildIntermediate(new ONearBlueprint(n.getDistance()), n); }
+
+    void visit(ProtonWeightedSetTerm &n) override { buildTerm(n); }
+    void visit(ProtonDotProduct &n)      override { buildTerm(n); }
+    void visit(ProtonWandTerm &n)        override { buildTerm(n); }
+
+    void visit(ProtonPhrase &n)          override { buildTerm(n); }
+    void visit(ProtonSameElement &n)     override { buildTerm(n); }
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/blueprintbuilder.cpp]: not generic term. need specific function with separate sub-visitor.

File: searchcore/src/vespa/searchcore/proton/matching/termdatafromnode.cpp [hunk 3]
```diff
@@ -19,29 +18,30 @@ struct TermDataFromTermVisitor
...
+    void visit(ProtonONear &) override {}
+    void visit(ProtonOr &) override {}
+    void visit(ProtonRank &) override {}
+    void visit(ProtonWeakAnd &) override {}
+
+    void visit(ProtonWeightedSetTerm &n) override { visitTerm(n); }
+    void visit(ProtonDotProduct &n) override { visitTerm(n); }
+    void visit(ProtonWandTerm &n) override { visitTerm(n); }
+    void visit(ProtonPhrase &n) override { visitTerm(n); }
+    void visit(ProtonSameElement &n) override { visitTerm(n); }
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/termdatafromnode.cpp]: the search iterator is (currently) being modeled without match data, which means that we either need to add match data to the iterator or remove the 'term-ness' from the query node.

File: searchlib/src/tests/query/querybuilder_test.cpp [hunk 4]
```diff
@@ -282,8 +271,7 @@ void checkQueryTreeTypes(Node *node) {
     string_term = dynamic_cast<StringTerm *>(wandTerm->getChildren()[1]);
     EXPECT_TRUE(checkTerm(string_term, str[2], view[2], id[2], weight[2]));
 
-    RegExpTerm *regexp_term =
-        dynamic_cast<RegExpTerm *>(and_node->getChildren()[8]);
+    RegExpTerm *regexp_term = dynamic_cast<RegExpTerm *>(and_node->getChildren()[8]);
```
Comment (havardpe) [inline searchlib/src/tests/query/querybuilder_test.cpp]: add testing of same element nodes here?

File: searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.cpp [hunk 5]
```diff
@@ -42,7 +42,13 @@ CreateBlueprintVisitorHelper::visitPhrase(search::query::Phrase &n) {
 }
 
 void
-CreateBlueprintVisitorHelper::handleNumberTermAsText(search::query::NumberTerm &n)
+CreateBlueprintVisitorHelper::visitSameElement(query::SameElement &n) {
+    (void) n;
+    //TODO Implement SameElementBlueprint creation.
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.cpp]: no, not here

File: searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.h [hunk 6]
```diff
@@ -37,42 +37,46 @@ class CreateBlueprintVisitorHelper : public search::query::QueryVisitor
...
-
-    void visit(search::query::NumberTerm &n) override = 0;
-    void visit(search::query::LocationTerm &n) override = 0;
-    void visit(search::query::PrefixTerm &n) override = 0;
-    void visit(search::query::RangeTerm &n) override = 0;
-    void visit(search::query::StringTerm &n) override = 0;
-    void visit(search::query::SubstringTerm &n) override = 0;
-    void visit(search::query::SuffixTerm &n) override = 0;
-    void visit(search::query::RegExpTerm &n) override = 0;
+    void visit(query::SameElement &n) override {
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.h]: should be handled by the parent visitor (query tree -> blueprint tree)

**Review by havardpe — COMMENTED** [review]
SameElement should probably be modeled more like Near and less like Phrase.

**Comment by baldersheim** [comment]
hmm, I think we should discuss this one further to ensure we get it right.

**Comment by MJPIPER99** [comment]
Please tell me what I will accomplish.

**Comment by baldersheim** [comment]
@havardpe Termlikeness removed PR again. Terms are also fully qualified from qrs.

File: searchcore/src/tests/proton/matching/querynodes_test.cpp [hunk 7]
```diff
@@ -210,9 +207,8 @@ class IteratorStructureTest {
 };
 
 typedef QueryBuilder<ProtonNodeTypes> QB;
-struct Phrase {
-    void addToBuilder(QB& b) { b.addPhrase(2, view, id, weight); }
-};
+struct Phrase { void addToBuilder(QB& b) { b.addPhrase(2, view, id, weight); }};
+struct SameElement { void addToBuilder(QB& b) { b.addSameElement(2, view); }};
```
Comment (havardpe) [inline searchcore/src/tests/proton/matching/querynodes_test.cpp]: should not have view

File: searchcorespi/src/vespa/searchcorespi/index/indexcollection.cpp (hunk 1 above)
Comment (havardpe) [reply searchcorespi/src/vespa/searchcorespi/index/indexcollection.cpp]: should still not be handled as a term in this context...

File: searchlib/src/tests/query/querybuilder_test.cpp (hunk 4 above)
Comment (havardpe) [reply searchlib/src/tests/query/querybuilder_test.cpp]: test still missing...

File: searchlib/src/tests/stackdumpiterator/stackdumpiteratortest.cpp [hunk 8]
```diff
@@ -244,6 +244,7 @@ StackDumpIteratorTest::RunTest(int testno, bool verify)
         stack.Push(new search::ParseItem(search::ParseItem::ITEM_NUMTERM, "foo", "[0;22]"));
         stack.Push(new search::ParseItem(search::ParseItem::ITEM_PREFIXTERM, "bar", "baz"));
         stack.Push(new search::ParseItem(search::ParseItem::ITEM_PHRASE, 3, "bar"));
+        stack.Push(new search::ParseItem(search::ParseItem::ITEM_SAME_ELEMENT, 3, "bar"));
```
Comment (havardpe) [inline searchlib/src/tests/stackdumpiterator/stackdumpiteratortest.cpp]: there should be no view here, this seems different from what I would expect the qrs to send us...

File: searchlib/src/vespa/searchlib/parsequery/parse.cpp [hunk 9]
```diff
@@ -111,8 +111,12 @@ ParseItem::AppendBuffer(RawBuf *buf) const
     case ITEM_WEAK_AND:
     case ITEM_NEAR:
     case ITEM_ONEAR:
+    case ITEM_SAME_ELEMENT:
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/parsequery/parse.cpp]: should move up to the other operations with the same payload (arity)

File: searchlib/src/vespa/searchlib/parsequery/parse.cpp [hunk 10]
```diff
@@ -199,6 +203,9 @@ ParseItem::GetBufferLen() const
     case ITEM_PHRASE:
         len += sizeof(uint32_t) * 2 + indexLen;
         break;
+    case ITEM_SAME_ELEMENT:
+        len += sizeof(uint32_t) * 2 + indexLen;
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/parsequery/parse.cpp]: should not contain view

File: searchlib/src/vespa/searchlib/parsequery/simplequerystack.cpp [hunk 11]
```diff
@@ -162,54 +163,58 @@ SimpleQueryStack::StackbufToString(const vespalib::stringref &theBuf)
...
             arity = tmp;
-            p += vespalib::compress::Integer::decompressPositive(tmp, p);
-            arg1 = tmp;
-            if (type == search::ParseItem::ITEM_WEAK_AND) {
+            if (type != ParseItem::ITEM_SAME_ELEMENT) {
+                p += vespalib::compress::Integer::decompressPositive(tmp, p);
+                arg1 = tmp;
+            }
+
+            if (type == ParseItem::ITEM_WEAK_AND || type == ParseItem::ITEM_SAME_ELEMENT) {
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/parsequery/simplequerystack.cpp]: decoding not matching encoding

File: searchlib/src/vespa/searchlib/parsequery/simplequerystack.cpp [hunk 12]
```diff
@@ -162,54 +163,58 @@ SimpleQueryStack::StackbufToString(const vespalib::stringref &theBuf)
...
+        case ParseItem::ITEM_ANY:
             p += vespalib::compress::Integer::decompressPositive(tmp, p);
             arity = tmp;
             result.append(make_string("%c/%d~", _G_ItemName[type], arity));
             break;
-        case search::ParseItem::ITEM_WEAK_AND:
-        case search::ParseItem::ITEM_NEAR:
-        case search::ParseItem::ITEM_ONEAR:
+        case ParseItem::ITEM_WEAK_AND:
+        case ParseItem::ITEM_SAME_ELEMENT:
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/parsequery/simplequerystack.cpp]: should move up to other items with only arity

File: searchlib/src/vespa/searchlib/parsequery/stackdumpiterator.cpp [hunk 13]
```diff
@@ -154,6 +152,20 @@ SimpleQueryStackDumpIterator::next()
         _currTerm = NULL;
         _currTermLen = 0;
         break;
+    case ParseItem::ITEM_SAME_ELEMENT:
+        if (p >= _bufEnd) return false;
+        p += vespalib::compress::Integer::decompressPositive(tmp, p);
+        _currArity = tmp;
+        _currArg1 = 0;
+        p += vespalib::compress::Integer::decompressPositive(tmp, p);
+        _currIndexNameLen = tmp;
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/parsequery/stackdumpiterator.cpp]: should not have view

File: searchlib/src/vespa/searchlib/query/query.cpp [hunk 14]
```diff
@@ -102,10 +102,11 @@ QueryConnector::create(ParseItem::ItemType type)
         case search::ParseItem::ITEM_WAND:         return new OrQueryNode();
         case search::ParseItem::ITEM_NOT:          return new AndNotQueryNode();
         case search::ParseItem::ITEM_PHRASE:       return new PhraseQueryNode();
+        case search::ParseItem::ITEM_SAME_ELEMENT: return new PhraseQueryNode(); // TODO: This needs a same element operation to work for streaming search too.
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/query/query.cpp]: AND would be a better fallback

File: searchlib/src/vespa/searchlib/query/tree/intermediatenodes.h [hunk 15]
```diff
@@ -105,6 +103,15 @@ class Phrase : public QueryNodeMixin<Phrase, Intermediate>, public Term {
     virtual ~Phrase() = 0;
 };
 
+class SameElement : public QueryNodeMixin<SameElement, Intermediate> {
+public:
+    SameElement(const vespalib::string &view) : _view(view) {}
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/query/tree/intermediatenodes.h]: should not have view

File: searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.cpp (hunk 5 above)
Comment (havardpe) [reply searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.cpp]: still not here ...

File: searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.h [hunk 16]
```diff
@@ -37,42 +37,46 @@ class CreateBlueprintVisitorHelper : public search::query::QueryVisitor
 
     const FieldSpec &getField() const { return _field; }
 
-    void visitPhrase(search::query::Phrase &n);
+    void visitPhrase(query::Phrase &n);
+    void visitSameElement(query::SameElement &n);
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.h]: not here

File: searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.h (hunk 6 above)
Comment (havardpe) [reply searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.h]: illegalVisit

**Review by havardpe — COMMENTED** [review]
please consider my comments
also, as far as I can tell, code is missing to avoid allocating match data for terms inside the SameElement operator

**Comment by havardpe** [comment]
I missed the fact that the qrs also sends the struct field name in addition to all mangled field names. Feel free to ignore my comments about not wanting a view for the same element operator (I can live with ignoring it), but make sure to align encoding/decoding.

File: searchcorespi/src/vespa/searchcorespi/index/indexcollection.cpp (hunk 1 above)
Comment (baldersheim) [reply searchcorespi/src/vespa/searchcorespi/index/indexcollection.cpp]: Correct, fixed

File: searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.h (hunk 6 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.h]: Done

File: searchlib/src/vespa/searchlib/parsequery/simplequerystack.cpp (hunk 11 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/parsequery/simplequerystack.cpp]: Code duplicated and for easier readability.

File: searchlib/src/vespa/searchlib/parsequery/simplequerystack.cpp (hunk 12 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/parsequery/simplequerystack.cpp]: Code duplicated and for easier readability.

**Review by baldersheim — COMMENTED** [review]
Have I messed up the review here, hmmm
Anyway I am keeping the view, until I figure out how to solve this for streaming search.
view is probably not the best name. Will contemplate once this is in.
If we find a simple way to populate the termnodes under the same node with full path we can avoid filling them in the qrs.

**Comment by baldersheim** [comment]
Merging to start writing system tests.

**Pull request merged**